reconstruct/reconstruct.go

```diff
@@ -326,31 +337,118 @@ func BubbleUp(node *expandedTree.ETree, prevNode *expandedTree.ETree, scrap []fl
+		log.Fatalf("Error: tree is not a well-formed binary tree. For node: %v, left is not nil but right is nil.\n", node.Name)
+	}
+	if node.Right != nil && node.Left == nil {
+		log.Fatalf("Error: tree is not a well-formed binary tree. For node: %v, right is not nil but left is nil.\n", node.Name)
+	}
+	if node.Left == nil && node.Right == nil { //if node is a leaf
+		// a descendent base is present at a leaf if the sequence at this position is not a gap.
+		node.DescendentBasePresent = node.Fasta.Seq[pos] != dna.Gap
+	} else { //if we are not a leaf
+		if node.Left != nil {
+			DescendentBaseExists(node.Left, pos)
+		}
+		if node.Right != nil {
+			DescendentBaseExists(node.Right, pos)
+		}
+		node.DescendentBasePresent = node.Left.DescendentBasePresent || node.Right.DescendentBasePresent
+	}
+}
+
+// LoopNodes performs ancestral sequence reconstruction for all nodes of an input tree, specified by an input root node,
+// at a user-specified alignment position.
+// Options: the user may specify a 'biasLeafName'. When specified, the reconstruction of this sequence's immediate ancestor
+// will be biased towards that descendent. The degree of this bias is controlled by the option 'nonBiasBaseThreshold'.
+// The user may also specify a 'highestProbThreshold'. If the program is uncertain about ancestral reconstruction for a particular
+// node, this option will allow LoopNodes to return an 'N' for that node instead.
+func LoopNodes(root *expandedTree.ETree, position int, biasLeafName string, nonBiasBaseThreshold float64, highestProbThreshold float64) {
+	var fix []float64
+	var biasBase, answerBase dna.Base
+	var biasParentName string
+	var biasLeafNode *expandedTree.ETree
+
+	if biasLeafName != "" {
+		biasLeafNode = expandedTree.FindNodeName(root, biasLeafName)
+		if biasLeafNode == nil {
+			log.Fatalf("Didn't find %v in tree.\n", biasLeafName)
+		}
+		if biasLeafNode.Up == nil {
+			log.Fatalf("Error: Bias reconstruction node was specified as the root node.")
+		}
+		biasParentName = biasLeafNode.Up.Name
+	}
+
 	internalNodes := expandedTree.GetBranch(root)
 	SetState(root, position)
-	for k := 0; k < len(internalNodes); k++ {
-		fix := FixFc(root, internalNodes[k])
-		yHat := Yhat(fix)
-		internalNodes[k].Fasta.Seq = append(internalNodes[k].Fasta.Seq, []dna.Base{dna.Base(yHat)}...)
+	DescendentBaseExists(root, position)
+	BaseExists(root, position)
+	for k := range internalNodes {
+		fix = FixFc(root, internalNodes[k])
+
+		if internalNodes[k].BasePresent {
+			if biasParentName != "" && internalNodes[k].Name == biasParentName {
+				biasBase = biasLeafNode.Fasta.Seq[position]
+				answerBase = LikelihoodsToBase(fix, nonBiasBaseThreshold, biasBase, highestProbThreshold) //biased estimate
+			} else {
+				answerBase = LikelihoodsToBase(fix, 0, dna.N, highestProbThreshold) //unbiased estimate
+			}
+		} else {
+			answerBase = dna.Gap
+		}
+
+		internalNodes[k].Fasta.Seq = append(internalNodes[k].Fasta.Seq, []dna.Base{answerBase}...)
```

Comment:
I think we are just adding one base to the slice, right?  So we can do:
internalNodes[k].Fasta.Seq = append(internalNodes[k].Fasta.Seq, answerBase)

Reply:
resolved in coming commit